### PR TITLE
HDDS-7398. Tool to remove old certs from the scm db

### PR DESCRIPTION
**Pull request by Galsza**
## What changes were proposed in this pull request?

Create a command line tool to remove expired certificates from the SCM metadata store.

## What is the link to the Apache JIRA

[HDDS-7398](https://issues.apache.org/jira/browse/HDDS-7398)

## How was this patch tested?

Added unit test for the command. There is also an ongoing test with a real cluster to see if the certs really get removed.

### CODE REVIEW

**Review by adoroszlai — COMMENTED** [review]
Thanks @Galsza for the patch.  Looks good overall.

Executing the following command (even on a datanode) returns success (rc=0), no error messages.

`ozone admin cert clean --db /tmp --name asdf`

I guess it should complain about not being able to open the DB.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/TypedTable.java [hunk 1]
```diff
@@ -34,8 +27,15 @@
...
-import org.apache.hadoop.hdds.utils.db.cache.TableCache.CacheType;
 import org.apache.hadoop.hdds.utils.db.cache.TableCache;
+import org.apache.hadoop.hdds.utils.db.cache.TableCache.CacheType;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
```
[the remark this answers is not in the document]
Comment (ChenSammi) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/TypedTable.java:38]: There is also imports moving in other java files.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/cert/CleanExpired.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import java.security.cert.X509Certificate;
+import java.time.Instant;
+import java.util.concurrent.Callable;
+
+/**
+ * This is the handler to clean SCM database from expired certificates.
+ */
+@CommandLine.Command(
+    name = "clean",
+    description = "Clean expired certificates from the SCM metadata",
```
Comment (ChenSammi) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/cert/CleanExpired.java:46]: Please add description about that this command only support to run when scm is shutdown.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/cert/CleanExpired.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    if (!Files.exists(Paths.get(dbDirectory))) {
+      System.out.println("DB path not exist:" + dbDirectory);
+      return null;
+    }
+
+    try (DBStore dbStore = HAUtils.loadDB(
+        configuration, new File(dbDirectory), dbName, new SCMDBDefinition())) {
+      removeExpiredCertificates(dbStore);
+    } catch (IOException e) {
+      System.out.println("Error trying to open" + dbName +
```
Comment (ChenSammi) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/cert/CleanExpired.java:80]: Please include the exception reason or message in the output error message.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/cert/CleanExpired.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        Table.KeyValue<?, ?> certPair = tableIterator.next();
+        X509Certificate certificate = (X509Certificate) certPair.getValue();
+        if (Instant.now().isAfter(certificate.getNotAfter().toInstant())) {
+          System.out.println("Certificate with id " + certPair.getKey() +
+              " and value: " + certificate + "will be deleted");
+          tableIterator.removeFromDB();
+        }
+      }
+    } catch (IOException e) {
+      System.out.println("Error when trying to open certificate table from db");
```
Comment (ChenSammi) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/cert/CleanExpired.java:103]: Please include the detail error cause.

**Comment by symious** [comment]
@Galsza Thanks for the work.

I was just wondering, is it also an option to add this feature as a daemon service in SCM, so that we don't have to do the cleaning manually?

**Comment by Galsza** [comment]
> @Galsza Thanks for the work.
> 
> I was just wondering, is it also an option to add this feature as a daemon service in SCM, so that we don't have to do the cleaning manually?

@symious Thanks for the comment. This is more of a housekeeping tool for older certificates. There is no need to add this feature as a daemon in SCM. The current plan for certificate rotation is that whenever a certificate is about the expire a new one will be created. This new certificate is used after instead of the old one. After some propagation time, when we can finally ensure that the old certificate is successfully replaced everywhere by the new one, the old certificate can be safely removed, which would be done there.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/metadata/SCMMetadataStoreImpl.java [hunk 5]
```diff
@@ -192,7 +191,7 @@ public void start(OzoneConfiguration config)
 
       metaTable = META.getTable(store);
 
-      checkAndPopulateTable(moveTable, META.getName());
+      checkAndPopulateTable(metaTable, META.getName());
```
Comment (fapifta) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/metadata/SCMMetadataStoreImpl.java:194]: Ouch...
Fortunately this affects just metrics as I see the tableMap that is populated here is not used elsewhere...

I believe this is irrelevant for the CLI tool, and this should be in a separate patch under a separate JIRA describing the problem we fix, and we should not hide it here.

**Review by fapifta — COMMENTED** [review]
Hi @Galsza thank you for taking this on, the patch overall seems to be good, and you had a very nice catch in the db definition. Let's extract that from this patch as it is irrelevant, and I am happy to +1 and commit this change.

As noted for that typo fix we should have a separate JIRA outlining the issue we had due to the typo ;)

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/metadata/SCMMetadataStoreImpl.java (hunk 5 above)
Comment (adoroszlai) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/metadata/SCMMetadataStoreImpl.java:194]: FYI, this is already fixed in HDDS-7490.

**Review by fapifta — APPROVED** [review]
Thank you for your work on this @Galsza the changes look good, and as I see all the comments were addressed.

@ChenSammi @adoroszlai  @symious please let me know if you have any further comments. if not, I will merge this before EOW.
